### DISCUSSION TITLE
자료구조 - 트리 - 호제&진호

### DISCUSSION THREAD

**Discussion by Jihn0118** [Shortudy-10th/tech-for-developer#58] [질의 응답]
- B트리와 B+트리의 구조와 그 차이점에 대해 설명해주세요.
- 레드-블랙트리에 대해 설명해주세요.
- AVL트리에 대해 해당 트리의 삽입, 검색, 삭제의 시간 복잡도를 포함하여 설명해주세요.
- 트리 순회를 위해 사용하는 방법에는 무엇이 있는지 설명해주세요
- BST(Binary Search Tree)와 Binary Tree에 대해 설명해주세요.
- 해당 트리의 높이와 차수를 말해주세요.
![tree](https://github.com/Shortudy-10th/tech-for-developer/assets/70586307/8e295580-04d5-47f9-b5c0-1b886cfacb5d)

**Comment by Jihn0118** [comment]
### ❓️ B트리와 B+트리의 구조와 그 차이점에 대해 설명해주세요

<br>

### 💬 답변
<blockquote>

![img1 daumcdn](https://github.com/Shortudy-10th/tech-for-developer/assets/83208807/f0d4b854-c2e2-4bc5-8dc9-e6e8c035248f)

B 트리는 데이터가 정렬된 상태로 유지되어 있는 균형 이진 트리의 확장판이다. 이진 트리와 유사하지만 한 노드 당 자식 노드가 2개 이상 가능하다.

B 트리의 장점은 어떤 값에 대해서도 같은 시간에 결과를 가질 수 있는 균일성을 가진다는 것이다.
선형 탐색의 경우 O(N)이 소요되지만 B트리에서는 O(logN)이 소요

이러한 B트리는 데이터 양이 증가함에 따라 효과가 상승한다.

![img1 daumcdn](https://github.com/Shortudy-10th/tech-for-developer/assets/83208807/b72ffda2-f204-493a-903d-07a2c4fd644b)

B+트리는 B 트리의 확장 개념으로 브랜치 노드에 key만 담아두고 data는 담지 않아, 오직 리프 노드에서만 key와 data를 저장하고 리프 노드끼리 연결 리스트로 연결되어 있다.

B+ 트리의 장점은 리프 노드를 제외하고 데이터를 담아두지 않기 때문에 메모리를 확보함으로써 더 많은 key를 수용할 수 있고, 하나의 노드에 더 많은 key들을 담을 수 있기에 트리의 높이가 낮아진다.
이는 곧 Cache hit을 높일 수 있음을 의미한다.


</blockquote>

**Reply by zini9188** [reply]
인덱싱에 b트리를 사용하는 이유를 알려주세요

1. 균형 트리로서 모든 리프 노드까지의 경로 길이가 같아 검색,삽입,삭제 작업의 시간 복잡도가 최대한 낮아지기 때문
2. 각 노드는 여러 개의 서브 트리를 가질 수 있으므로 높은 브랜치팩터를 활용할 수 있어 트리의 높이가 줄어들고 디스크 I/O 연산 비용이 절약되기 때문
3. 새로운 데이터가 추가되거나 삭제되어도 곧바로 재구조화되어 균형을 유지할 수 있는 적응형 자료구조이기 때문
4. 데이터베이스 파일 시스템 또는 메모리에 적용할 수 있는 효과적인 인덱스 구조이기 때문

**Comment by Jihn0118** [comment]
### ❓️ 레드-블랙트리에 대해 설명해주세요.

<br>

### 💬 답변
<blockquote>


![img1 daumcdn](https://github.com/Shortudy-10th/tech-for-developer/assets/83208807/4ede30a1-1960-4ac7-9974-86c01f60aa07)

자가 균형 이진 탐색 트리로 연관 배열을 구현하는 데 쓰이는 자료구조
특정 조건을 지키며 최악의 경우에도 O(logN) 시간 안에 가능하도록 해준다.

### 조건
- 모든 노드가 RED이거나 BLACK이다.
- 루트 노드는 BLACK이다.
- 모든 리프노드는 BLACK이다.
- 노드가 RED이면 그 자식은 BLACK이다.
  - RED 노드는 연속으로 나올 수 없다.
- 루트 노드에서 모든 리프 노드까지의 경로에서 만나는 BLACK 노드의 수는 같다.

### 장점
- 최악의 경우에도 일정한 실행 시간을 보장
- 실시간 처리와 같은 실행 시간이 중요한 경우에 유용하게 사용

### 단점

- 이해하기 어렵고 구현이 복잡하다.



</blockquote>

<br>

**Comment by Jihn0118** [comment]
### ❓️ AVL트리에 대해 해당 트리의 삽입, 검색, 삭제의 시간 복잡도를 포함하여 설명해주세요.

<br>

### 💬 답변
<blockquote>

AVL트리: 이진탐색트리에서 한 쪽으로 노드가 쏠린 트리 형태의 단점을 보완한 **자가 균형 이진 탐색 트리**의 일종으로 트리의 높이를 항상 균형 있게 유지하여 검색, 삽입, 삭제 연산의 평균적인 시간 복잡도를 **O(log n)** 으로 유지한다.

AVL 트리 시뮬레이터
https://www.cs.usfca.edu/~galles/visualization/AVLtree.html

</blockquote>

<br>

**Reply by Jihn0118** [reply]
### ❓️ 어떻게 AVL트리는 높이를 항상 균형 있게 유지하나요?

<br>

### 💬 답변
<blockquote>

AVL트리는 왼쪽, 오른쪽 서브 트리의 높이 차이가 최대 1인데, 높이 차이가 1보다 커지면 회전(Rotation)을 통해 균형을 맞춰 높이 차이를 줄인다.

AVL트리는 균형이 무너졌는 지에 대해 판단할 때 Balance Factor라는 것을 활용한다.
BF(K) = K의 왼쪽 서브 트리의 높이 - K의 오른쪽 서브트리
이를 벗어나면 균형이 깨졌다는 것을 의미하고, 이때 회전이 필요한 것이다. 

특정 노드의 삽입 또는 삭제를 할 때 먼저 일반적인 BST의 삽입, 삭제 방법을 따르는데, 
삽입 or 삭제 후에 균형 유지를 위해 회전 연산을 한다. 
검색 연산의 시간 복잡도는 트리의 높이에 비례하게 되는데, 균형 있는 트리이므로 시간 복잡도는 O(log n)이다. 

회전 연산의 시간 복잡도는 O(1)이기 때문에 검색, 삽입, 검색의 총 시간 복잡도는 O(log n)이 된다.
</blockquote>

참조: https://code-lab1.tistory.com/61

**Comment by Jihn0118** [comment]
### ❓️ 트리 순회를 위해 사용하는 방법에는 무엇이 있는지 설명해주세요.

<br>

### 💬 답변
<blockquote>
순회: 트리의 노드들을 체계적으로 방문하는 것


1. 전위순회(Preorder Traversal)
  부모 노드 방문 후, 자식 노드를 좌우 순서로 방문한다.
2. 중위순회(Inorder Traversal)
  왼쪽 자식 노드, 부모 노드, 오른쪽 자식 노드 순으로 방문한다.
3. 후위순회(Postorder Traversal)
  자식 노드를 좌우 순서로 방문한 후, 부모 노드로 방문한다.
</blockquote>

**Comment by Jihn0118** [comment]
### ❓️ 주어진 트리의 높이와 차수를 말해주세요.
![tree](https://github.com/Shortudy-10th/tech-for-developer/assets/70586307/8e295580-04d5-47f9-b5c0-1b886cfacb5d)

<br>

### 💬 답변
<blockquote>

트리의 차수: 트리에 있는 노드의 차수 중에서 가장 큰 값
- 이 트리에서 차수 - **3** 

트리의 높이: 트리에 있는 노드의 높이 중에서 가장 큰 값. 최대 레벨
- 이 트리에서 높이 - **5** 

</blockquote>

<br>

**Comment by Jihn0118** [comment]
### ❓️ BST(Binary Search Tree)와 Binary Tree에 대해 설명해주세요.

<br>

### 💬 답변
<blockquote>
이진 트리(Binary Tree): 모든 노드들이  둘 이하의 자식을 가진 트리

이진 트리를 배열로 표현하게 되면 아래와 같은 형태가 된다.

![img](https://github.com/Shortudy-10th/tech-for-developer/assets/87293724/1ed72d53-2022-495f-84b3-67279c42f1cd)


이진 탐색 트리(Binary Search Tree): 정렬된 이진트리

![1024px-Binary_search_tree svg](https://github.com/Shortudy-10th/tech-for-developer/assets/87293724/f4d8b46f-a891-4068-9fb1-8966ca1591dd)

<특징> 
- 노드의 왼쪽 서브 트리에는 그 노드의 값보다 작은 값들을 지닌 노드들로 이루어져 있다.
-  노드의 오른쪽 서브 트리에는 그 노드의 값보다 큰 값들을 지닌 노드들로 이루어져 있다.
- 좌우 서브 트리는 각각 다시 이진 탐색 트리여야 한다.


</blockquote>

<br>
